pkg/services/network/nsxvpc_provider.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	nsxopv1 "github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	vmoprv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	vmwarev1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/vmware/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/context/vmware"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/services"
+)
+
+// nsxvpcNetworkProvider provisions nsx-vpc type cluster network.
+type nsxvpcNetworkProvider struct {
+	client client.Client
+}
+
+// NsxVpcNetworkProvider returns an instance of nsx-vpc type network provider.
+func NsxVpcNetworkProvider(client client.Client) services.NetworkProvider {
+	return &nsxvpcNetworkProvider{
+		client: client,
+	}
+}
+
+func (vp *nsxvpcNetworkProvider) HasLoadBalancer() bool {
+	return true
+}
+
+func (vp *nsxvpcNetworkProvider) SupportsVMRedinessProbe() bool {
+	// Note: The control plane VM network is private with nsx-vpc and
+	// readiness probe would fail. Therefore, nsxvpcNetworkProvider
+	// doesn't support VM readiness probe.
+	return false
+}
+
+// GetNsxVpcSubnetSetName returns the name of the NSX VPC SubnetSet object.
+func GetNsxVpcSubnetSetName(clusterName string) string {
+	return fmt.Sprintf("%s-subnetset", clusterName)
+}
+
+// verifyNsxVpcSubnetSetStatus checks the status conditions of a given SubnetSet within a cluster context.
+// If the subnet isn't ready, it is marked as false, and the function returns an error.
+// If the subnet is ready, the function updates the VSphereCluster with a "true" status and returns nil.
+func (vp *nsxvpcNetworkProvider) verifyNsxVpcSubnetSetStatus(ctx *vmware.ClusterContext, subnetset *nsxopv1.SubnetSet) error {
```

Comment:
I'd prefer not having `vmware.ClusterContext` as parameter here (I think only VSphereCluster is used?), please pass down instead what really is requred.

also not name it ctx, `ctx` should only be used for real `context.Context` objects to not confuse when reading code :-)

Reply:
Will address it.